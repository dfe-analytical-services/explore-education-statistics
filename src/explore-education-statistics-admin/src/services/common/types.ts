export interface IdTitlePair {
  id: string;
  title: string;
}

export interface IdLabelPair {
  id: string;
  label: string;
}

export interface ValueLabelPair {
  value: string;
  label: string;
}

export interface ContactDetails {
  id: string;
  contactName: string;
  contactTelNo: string;
  teamEmail: string;
  teamName?: string;
}

export interface UserDetails {
  id: string;
  name: string;
}

export interface TimePeriodCoverageGroup {
  category: {
    label: string;
  };
  timeIdentifiers: {
    identifier: ValueLabelPair;
  }[];
}

export interface BasicPublicationDetails {
  id: string;
  title: string;
  contact?: ContactDetails;
  methodologyId?: string;
  themeId: string;
}
