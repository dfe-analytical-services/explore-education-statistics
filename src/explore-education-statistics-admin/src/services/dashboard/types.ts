import {
  ContactDetails,
  IdLabelPair,
  IdTitlePair,
  UserDetails,
} from '@admin/services/common/types';
import {
  DayMonthYearValues,
  ReleaseStatus,
} from '@common/services/publicationService';

export interface ThemeAndTopics {
  title: string;
  id: string;
  topics: {
    title: string;
    id: string;
  }[];
}

export interface Comment {
  message: string;
  authorName: string;
  createdDate: string;
}

export interface AdminDashboardRelease {
  id: string;
  status: ReleaseStatus;
  latestRelease: boolean;
  live: boolean;
  releaseName: string;
  publicationId: string;
  publicationTitle: string;
  timePeriodCoverage: IdLabelPair;
  contact: ContactDetails;
  lastEditedUser: UserDetails;
  lastEditedDateTime: string;
  publishScheduled: Date;
  published?: string;
  nextReleaseDate: DayMonthYearValues;
  internalReleaseNote?: string;
  draftComments: Comment[];
  higherReviewComments: Comment[];
}

export interface AdminDashboardPublication {
  id: string;
  title: string;
  methodology?: IdTitlePair;
  releases: AdminDashboardRelease[];
  contact: ContactDetails;
}
