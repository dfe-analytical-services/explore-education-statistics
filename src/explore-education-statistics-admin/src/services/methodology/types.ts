export interface MethodologyStatus {
  id: string;
  title: string;
  status: string;
  publications: MethodologyStatusPublication[];
}

export interface MethodologyStatusPublication {
  id: string;
  title: string;
}
