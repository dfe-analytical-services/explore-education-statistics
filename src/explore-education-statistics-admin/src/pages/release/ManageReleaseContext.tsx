import { BasicPublicationDetails } from '@admin/services/common/types';
import { createContext, useContext } from 'react';

export interface ManageRelease {
  publication: BasicPublicationDetails;
  releaseId: string;
  onChangeReleaseStatus: () => void;
}

const ManageReleaseContext = createContext<ManageRelease | undefined>(
  undefined,
);

export function useManageReleaseContext() {
  const context = useContext(ManageReleaseContext);

  if (context === undefined) {
    throw new Error(
      'useManageReleaseContext must be used within a ManageReleaseContext',
    );
  }
  return context;
}

export default ManageReleaseContext;
