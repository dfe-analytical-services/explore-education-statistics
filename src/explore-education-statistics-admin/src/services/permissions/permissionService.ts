import { GlobalPermissions, User } from '@admin/services/sign-in/types';
import client from '@admin/services/util/service';

export type PreReleaseAccess = 'Before' | 'After' | 'Within' | 'NoneSet';

export interface PreReleaseWindowStatus {
  preReleaseAccess: PreReleaseAccess;
  preReleaseWindowStartTime: Date;
  preReleaseWindowEndTime: Date;
}

const permissionService = {
  getGlobalPermissions: (): Promise<GlobalPermissions> => {
    return client.get(`/permissions/access`);
  },
  canAccessPrereleasePages: (user?: User): Promise<boolean> => {
    return Promise.resolve(
      user ? user.permissions.canAccessPrereleasePages : false,
    );
  },
  canUpdateRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/update`);
  },
  // BAU-324 - temporary stopgap to stop the updating of Release Amendment data files until Phase 2 of
  // Release Versioning is tackled
  canUpdateReleaseDataFiles: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/update-data-files`);
  },
  canMarkReleaseAsDraft: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/draft`);
  },
  canSubmitReleaseForHigherLevelReview: (
    releaseId: string,
  ): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/submit`);
  },
  canApproveRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/approve`);
  },
  canMakeAmendmentOfRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/amend`);
  },
  canCreatePublicationForTopic: (topicId: string): Promise<boolean> => {
    return client.get(`/permissions/topic/${topicId}/publication/create`);
  },
  canCreateReleaseForPublication: (publicationId: string): Promise<boolean> => {
    return client.get(
      `/permissions/publication/${publicationId}/release/create`,
    );
  },
  canMarkMethodologyAsDraft: (methodologyId: string): Promise<boolean> => {
    return client.get(`/permissions/methodology/${methodologyId}/status/draft`);
  },
  canApproveMethodology: (methodologyId: string): Promise<boolean> => {
    return client.get(
      `/permissions/methodology/${methodologyId}/status/approve`,
    );
  },
  getPreReleaseWindowStatus: (
    releaseId: string,
  ): Promise<PreReleaseWindowStatus> => {
    return client
      .get<PreReleaseWindowStatus>(
        `/permissions/release/${releaseId}/prerelease/status`,
      )
      .then(status => ({
        ...status,
        preReleaseWindowStartTime: new Date(status.preReleaseWindowStartTime),
        preReleaseWindowEndTime: new Date(status.preReleaseWindowEndTime),
      }));
  },
};

export default permissionService;
