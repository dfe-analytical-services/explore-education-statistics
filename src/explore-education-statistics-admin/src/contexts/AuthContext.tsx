import authService from '@admin/components/api-authorization/AuthorizeService';
import permissionService, {
  GlobalPermissions,
} from '@admin/services/permissionService';
import logger from '@common/services/logger';
import React, {
  createContext,
  ReactElement,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

export interface User {
  id: string;
  name: string;
  validToken?: boolean;
  permissions: GlobalPermissions;
}

const defaultPermissions: GlobalPermissions = {
  canAccessSystem: false,
  canAccessPrereleasePages: false,
  canAccessAnalystPages: false,
  canAccessAllImports: false,
  canManageAllTaxonomy: false,
  isBauUser: false,
};

export interface AuthContextState {
  user?: User;
}

export const AuthContext = createContext<AuthContextState | undefined>(
  undefined,
);

interface Props {
  children: ReactNode;
}

interface State {
  ready: boolean;
  user?: User;
}

export const AuthContextProvider = ({ children }: Props) => {
  const [state, setState] = useState<State>({ ready: false });

  useEffect(() => {
    const populateAuthenticationState = async () => {
      const authenticated = await authService.isAuthenticated();

      if (authenticated) {
        try {
          const userProfile = await authService.getUser();

          if (!userProfile) {
            setState({
              ready: true,
            });

            return;
          }

          const { profile } = userProfile;
          const userId = profile.sub;

          const validToken =
            new Date() < new Date(userProfile.expires_at * 1000);

          const permissions = validToken
            ? await permissionService.getGlobalPermissions()
            : defaultPermissions;

          setState({
            ready: true,
            user: {
              id: userId,
              name: profile.given_name ?? '',
              validToken,
              permissions,
            },
          });
        } catch (err) {
          logger.error(err);

          setState({ ready: false });
        }
      } else {
        setState({ ready: true });
      }
    };

    let subscriptionId: number;

    populateAuthenticationState().then(() => {
      subscriptionId = authService.subscribe(async () => {
        await populateAuthenticationState();
      });
    });

    return () => {
      if (typeof subscriptionId !== 'undefined') {
        authService.unsubscribe(subscriptionId);
      }
    };
  }, []);

  const contextState: AuthContextState = useMemo(() => {
    return {
      user: state.user,
    };
  }, [state.user]);

  return state.ready ? (
    <AuthContext.Provider value={contextState}>{children}</AuthContext.Provider>
  ) : null;
};

export function useAuthContext(): AuthContextState {
  const context = useContext(AuthContext);
  return context ?? {};
}

interface AuthContextTestProviderProps {
  children: ReactNode;
  user: User;
}

export function AuthContextTestProvider({
  children,
  user,
}: AuthContextTestProviderProps): ReactElement {
  return (
    // eslint-disable-next-line react/jsx-no-constructed-context-values
    <AuthContext.Provider value={{ user }}>{children}</AuthContext.Provider>
  );
}
