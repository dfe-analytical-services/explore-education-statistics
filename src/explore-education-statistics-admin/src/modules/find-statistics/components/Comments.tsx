import { useAuthContext } from '@admin/contexts/AuthContext';
import { useManageReleaseContext } from '@admin/pages/release/ManageReleaseContext';
import releaseContentCommentService, {
  AddComment,
  UpdateComment,
} from '@admin/services/releaseContentCommentService';
import { Comment } from '@admin/services/types/content';
import Button from '@common/components/Button';
import ButtonGroup from '@common/components/ButtonGroup';
import ButtonText from '@common/components/ButtonText';
import Details from '@common/components/Details';
import { Form } from '@common/components/form';
import FormFieldTextArea from '@common/components/form/FormFieldTextArea';
import FormattedDate from '@common/components/FormattedDate';
import Yup from '@common/validation/yup';
import { Formik } from 'formik';
import React, { useState } from 'react';
import styles from './Comments.module.scss';

interface FormValues {
  content: string;
}

export type CommentsChangeHandler = (
  blockId: string,
  comments: Comment[],
) => void;

interface Props {
  blockId: string;
  sectionId: string;
  comments: Comment[];
  canComment?: boolean;
  onChange: CommentsChangeHandler;
}

const Comments = ({
  blockId,
  sectionId,
  comments = [],
  onChange,
  canComment = true,
}: Props) => {
  const [editingComment, setEditingComment] = useState<Comment>();

  const { user } = useAuthContext();
  const { releaseId } = useManageReleaseContext();

  const addComment = (content: string) => {
    const additionalComment: AddComment = {
      content,
    };

    if (releaseId && sectionId) {
      releaseContentCommentService
        .addContentSectionComment(
          releaseId,
          sectionId,
          blockId,
          additionalComment,
        )
        .then(newComment => {
          onChange(blockId, [newComment, ...comments]);
        });
    }
  };

  const removeComment = (index: number) => {
    const commentId = comments[index].id;

    releaseContentCommentService
      .deleteContentSectionComment(commentId)
      .then(() => {
        const newComments = [...comments];
        newComments.splice(index, 1);

        onChange(blockId, newComments);
      });
  };

  const updateComment = (index: number, content: string) => {
    const editedComment: UpdateComment = {
      ...comments[index],
      content,
    };

    releaseContentCommentService
      .updateContentSectionComment(editedComment)
      .then(savedComment => {
        const newComments = [...comments];
        newComments[index] = savedComment;

        onChange(blockId, newComments);

        setEditingComment(undefined);
      });
  };

  return (
    <div className={styles.container}>
      <Details
        className="govuk-!-margin-bottom-1 govuk-body-s"
        summary={`${canComment ? `Add / ` : ''}View comments (${
          comments.length
        })`}
      >
        {canComment && (
          <Formik<FormValues>
            initialValues={{
              content: '',
            }}
            validationSchema={Yup.object({
              content: Yup.string().required('Enter a comment'),
            })}
            onSubmit={values => {
              addComment(values.content);
            }}
          >
            <Form id={`${blockId}-addCommentForm`} showErrorSummary={false}>
              <FormFieldTextArea<FormValues>
                label="Comment"
                name="content"
                id={`${blockId}-addCommentForm-content`}
                rows={3}
              />

              <Button type="submit">Add comment</Button>
            </Form>
          </Formik>
        )}

        <ul className={styles.commentsContainer}>
          {comments.map((comment, index) => {
            const { createdBy } = comment;

            return (
              <li key={comment.id} className="govuk-body-s">
                <p className="govuk-!-margin-0">
                  <strong>
                    {`${createdBy.firstName} ${createdBy.lastName}`}
                  </strong>
                </p>
                <p>
                  {comment.updated ? 'Updated: ' : ''}
                  <FormattedDate format="d MMMM yyyy HH:mm">
                    {comment.updated || comment.created}
                  </FormattedDate>
                </p>

                {editingComment?.id === comment.id ? (
                  <Formik<FormValues>
                    initialValues={{
                      content: editingComment.content,
                    }}
                    validationSchema={Yup.object({
                      content: Yup.string().required('Enter a comment'),
                    })}
                    onSubmit={values => {
                      updateComment(index, values.content);
                    }}
                  >
                    <Form
                      id={`${blockId}-editCommentForm`}
                      showErrorSummary={false}
                    >
                      <FormFieldTextArea<FormValues>
                        label="Comment"
                        name="content"
                        id={`${blockId}-editCommentForm-editComment`}
                        rows={3}
                      />

                      <ButtonGroup>
                        <Button type="submit">Update</Button>
                        <ButtonText
                          onClick={() => {
                            setEditingComment(undefined);
                          }}
                        >
                          Cancel
                        </ButtonText>
                      </ButtonGroup>
                    </Form>
                  </Formik>
                ) : (
                  <>
                    <p>{comment.content}</p>

                    {canComment && user?.id === createdBy.id && (
                      <ButtonGroup>
                        <ButtonText
                          onClick={() => {
                            setEditingComment(comment);
                          }}
                        >
                          Edit
                        </ButtonText>

                        <ButtonText
                          onClick={() => {
                            removeComment(index);
                          }}
                        >
                          Delete
                        </ButtonText>
                      </ButtonGroup>
                    )}
                  </>
                )}

                <hr />
              </li>
            );
          })}
        </ul>
      </Details>
    </div>
  );
};

export default Comments;
