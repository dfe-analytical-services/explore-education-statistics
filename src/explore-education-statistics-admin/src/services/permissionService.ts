import { User } from '@admin/contexts/AuthContext';
import client from '@admin/services/utils/service';

export type PreReleaseAccess = 'Before' | 'After' | 'Within' | 'NoneSet';

export interface PreReleaseWindowStatus {
  access: PreReleaseAccess;
  start: Date;
  end: Date;
}

export interface GlobalPermissions {
  canAccessSystem: boolean;
  canAccessPrereleasePages: boolean;
  canAccessAnalystPages: boolean;
  canAccessUserAdministrationPages: boolean;
  canAccessMethodologyAdministrationPages: boolean;
}

const permissionService = {
  getGlobalPermissions: (): Promise<GlobalPermissions> => {
    return client.get(`/permissions/access`);
  },
  canAccessPrereleasePages: (user?: User): Promise<boolean> => {
    return Promise.resolve(
      user ? user.permissions.canAccessPrereleasePages : false,
    );
  },
  canUpdateRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/update`);
  },
  canMarkReleaseAsDraft: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/draft`);
  },
  canSubmitReleaseForHigherLevelReview: (
    releaseId: string,
  ): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/submit`);
  },
  canApproveRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/status/approve`);
  },
  canMakeAmendmentOfRelease: (releaseId: string): Promise<boolean> => {
    return client.get(`/permissions/release/${releaseId}/amend`);
  },
  canCreatePublicationForTopic: (topicId: string): Promise<boolean> => {
    return client.get(`/permissions/topic/${topicId}/publication/create`);
  },
  canCreateReleaseForPublication: (publicationId: string): Promise<boolean> => {
    return client.get(
      `/permissions/publication/${publicationId}/release/create`,
    );
  },
  canUpdateMethodology: (methodologyId: string): Promise<boolean> => {
    return client.get(`/permissions/methodology/${methodologyId}/update`);
  },
  canMarkMethodologyAsDraft: (methodologyId: string): Promise<boolean> => {
    return client.get(`/permissions/methodology/${methodologyId}/status/draft`);
  },
  canApproveMethodology: (methodologyId: string): Promise<boolean> => {
    return client.get(
      `/permissions/methodology/${methodologyId}/status/approve`,
    );
  },
  getPreReleaseWindowStatus: (
    releaseId: string,
  ): Promise<PreReleaseWindowStatus> => {
    return client
      .get<{
        access: PreReleaseAccess;
        start: string;
        end: string;
      }>(`/permissions/release/${releaseId}/prerelease/status`)
      .then(status => ({
        access: status.access,
        start: new Date(status.start),
        end: new Date(status.end),
      }));
  },
};

export default permissionService;
