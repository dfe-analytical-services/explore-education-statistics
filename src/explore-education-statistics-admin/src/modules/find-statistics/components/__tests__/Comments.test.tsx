import { AuthContext, User } from '@admin/contexts/AuthContext';
import ManageReleaseContext from '@admin/pages/release/ManageReleaseContext';
import { GlobalPermissions } from '@admin/services/permissionService';
import _releaseContentCommentService from '@admin/services/releaseContentCommentService';
import { Comment } from '@admin/services/types/content';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import noop from 'lodash/noop';
import React from 'react';
import Comments from '../Comments';

jest.mock('@admin/services/releaseContentCommentService');

const releaseContentCommentService = _releaseContentCommentService as jest.Mocked<
  typeof _releaseContentCommentService
>;

describe('Comments', () => {
  const testRelease = {
    releaseId: 'release-1',
    onChangeReleaseStatus: noop,
    publication: {
      id: 'publication-1',
      themeId: 'theme-1',
      title: 'Test publication',
    },
  };

  test('renders list of comments when comments are opened', () => {
    render(
      <ManageReleaseContext.Provider value={testRelease}>
        <Comments
          blockId="test-block"
          sectionId="section-1"
          comments={[
            {
              id: 'comment-1',
              content: 'Test comment 1',
              createdBy: {
                id: 'user-1',
                email: 'test@test.com',
                firstName: 'John',
                lastName: 'Smith',
              },
              created: '2020-06-06T12:00:00',
            },
            {
              id: 'comment-2',
              content: 'Test comment 2',
              createdBy: {
                id: 'user-1',
                email: 'test2@test.com',
                firstName: 'Jane',
                lastName: 'Roberts',
              },
              created: '2020-06-06T11:00:00',
              updated: '2020-06-07T15:00:00',
            },
          ]}
          onChange={noop}
        />
      </ManageReleaseContext.Provider>,
    );

    fireEvent.click(
      screen.getByRole('button', {
        name: 'Add / View comments (2)',
      }),
    );

    const comments = screen.getAllByRole('listitem');

    expect(comments).toHaveLength(2);

    expect(comments[0]).toHaveTextContent('John Smith');
    expect(comments[0]).toHaveTextContent('6 June 2020');
    expect(comments[0]).toHaveTextContent('Test comment 1');

    expect(comments[1]).toHaveTextContent('Jane Roberts');
    expect(comments[1]).toHaveTextContent('Updated: 7 June 2020');
    expect(comments[1]).toHaveTextContent('Test comment 2');

    expect(screen.getByRole('list')).toMatchSnapshot();
  });

  test('adding comment calls the `onChange` handler with new comments list', async () => {
    const handleChange = jest.fn();

    render(
      <ManageReleaseContext.Provider value={testRelease}>
        <Comments
          blockId="test-block"
          sectionId="section-1"
          comments={[
            {
              id: 'comment-1',
              content: 'Test comment 1',
              createdBy: {
                id: 'user-1',
                email: 'test@test.com',
                firstName: 'John',
                lastName: 'Smith',
              },
              created: '2020-06-06T12:00:00',
            },
          ]}
          onChange={handleChange}
        />
      </ManageReleaseContext.Provider>,
    );

    fireEvent.change(
      screen.getByRole('textbox', {
        name: 'Comment',
        hidden: true,
      }),
      {
        target: {
          value: 'New test comment',
        },
      },
    );

    releaseContentCommentService.addContentSectionComment.mockImplementation(
      () =>
        Promise.resolve<Comment>({
          id: 'comment-2',
          content: 'New test comment',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Bethany',
            lastName: 'Parker',
          },
          created: '2020-06-06T14:00:00',
        }),
    );

    fireEvent.click(
      screen.getByRole('button', {
        name: 'Add comment',
        hidden: true,
      }),
    );

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledWith('test-block', [
        {
          id: 'comment-2',
          content: 'New test comment',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Bethany',
            lastName: 'Parker',
          },
          created: '2020-06-06T14:00:00',
        },
        {
          id: 'comment-1',
          content: 'Test comment 1',
          createdBy: {
            id: 'user-1',
            email: 'test@test.com',
            firstName: 'John',
            lastName: 'Smith',
          },
          created: '2020-06-06T12:00:00',
        },
      ] as Comment[]);
    });
  });

  describe('updating comment', () => {
    const testComments = [
      {
        id: 'comment-1',
        content: 'Test comment 1',
        createdBy: {
          id: 'user-1',
          email: 'test@test.com',
          firstName: 'John',
          lastName: 'Smith',
        },
        created: '2020-06-06T12:00:00',
      },
      {
        id: 'comment-2',
        content: 'Test comment 2',
        createdBy: {
          id: 'user-2',
          email: 'test2@test.com',
          firstName: 'Jane',
          lastName: 'Roberts',
        },
        created: '2020-06-06T11:00:00',
      },
    ];

    const testUser: User = {
      id: 'user-2',
      name: 'Jane Roberts',
      permissions: {} as GlobalPermissions,
    };

    test('can only update comment belonging to the current user', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeNull();

      expect(
        within(comments[1]).getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('clicking `Edit` button shows textbox and hides `Edit` button', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      fireEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      expect(
        comment.getByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
      ).toBeInTheDocument();

      expect(
        comment.queryByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).not.toBeInTheDocument();
    });

    test('clicking `Cancel` button hides the textbox again', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      fireEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      fireEvent.click(
        comment.getByRole('button', {
          name: 'Cancel',
          hidden: true,
        }),
      );

      expect(
        comment.queryByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
      ).not.toBeInTheDocument();

      expect(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('updating comment calls the `onChange` handler with the updated comment', async () => {
      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      fireEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      fireEvent.change(
        comment.getByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
        {
          target: {
            value: 'Updated test comment',
          },
        },
      );

      releaseContentCommentService.updateContentSectionComment.mockImplementation(
        () =>
          Promise.resolve<Comment>({
            id: 'comment-2',
            content: 'Updated test comment',
            createdBy: {
              id: 'user-2',
              email: 'test2@test.com',
              firstName: 'Jane',
              lastName: 'Roberts',
            },
            created: '2020-06-06T11:00:00',
            updated: '2020-06-06T15:00:00',
          }),
      );

      fireEvent.click(
        comment.getByRole('button', {
          name: 'Update',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          {
            id: 'comment-1',
            content: 'Test comment 1',
            createdBy: {
              id: 'user-1',
              email: 'test@test.com',
              firstName: 'John',
              lastName: 'Smith',
            },
            created: '2020-06-06T12:00:00',
          },
          {
            id: 'comment-2',
            content: 'Updated test comment',
            createdBy: {
              id: 'user-2',
              email: 'test2@test.com',
              firstName: 'Jane',
              lastName: 'Roberts',
            },
            created: '2020-06-06T11:00:00',
            updated: '2020-06-06T15:00:00',
          },
        ] as Comment[]);
      });
    });
  });

  describe('deleting comment', () => {
    const testComments: Comment[] = [
      {
        id: 'comment-1',
        content: 'Test comment 1',
        createdBy: {
          id: 'user-1',
          email: 'test@test.com',
          firstName: 'John',
          lastName: 'Smith',
        },
        created: '2020-06-06T12:00:00',
      },
      {
        id: 'comment-2',
        content: 'Test comment 2',
        createdBy: {
          id: 'user-2',
          email: 'test2@test.com',
          firstName: 'Jane',
          lastName: 'Roberts',
        },
        created: '2020-06-06T11:00:00',
      },
    ];

    const testUser: User = {
      id: 'user-2',
      name: 'Jane Roberts',
      permissions: {} as GlobalPermissions,
    };

    test('can only delete comment belonging to the current user', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      ).toBeNull();

      expect(
        within(comments[1]).getByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('deleting comment calls the `onChange` handler with it removed from comments', async () => {
      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ManageReleaseContext.Provider value={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ManageReleaseContext.Provider>
        </AuthContext.Provider>,
      );

      releaseContentCommentService.deleteContentSectionComment.mockImplementation(
        () => Promise.resolve(),
      );

      fireEvent.click(
        screen.getByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          {
            id: 'comment-1',
            content: 'Test comment 1',
            createdBy: {
              id: 'user-1',
              email: 'test@test.com',
              firstName: 'John',
              lastName: 'Smith',
            },
            created: '2020-06-06T12:00:00',
          },
        ] as Comment[]);
      });
    });
  });
});
