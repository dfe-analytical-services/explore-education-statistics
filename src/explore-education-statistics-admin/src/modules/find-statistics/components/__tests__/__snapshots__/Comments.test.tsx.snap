// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments renders list of comments when comments are opened 1`] = `
<ul
  class="commentsContainer"
>
  <li
    class="govuk-body-s"
  >
    <p
      class="govuk-!-margin-0"
    >
      <strong>
        John Smith
      </strong>
    </p>
    <p>
      
      <time>
        6 June 2020 12:00
      </time>
    </p>
    <p>
      Test comment 1
    </p>
    <hr />
  </li>
  <li
    class="govuk-body-s"
  >
    <p
      class="govuk-!-margin-0"
    >
      <strong>
        Jane Roberts
      </strong>
    </p>
    <p>
      Updated: 
      <time>
        7 June 2020 15:00
      </time>
    </p>
    <p>
      Test comment 2
    </p>
    <hr />
  </li>
</ul>
`;
